File: Residential_Controller.py
import time

class elevator: 
    floorList = []
    def __init__(self, status, direction, floorNumber, elevatorNumber):
       self.status = status
       self.direction = direction
       self.floorNumber = floorNumber
       self.elevatorNumber = elevatorNumber
    
    def addToFloorList(self, floorNumber):
        self.floorList.append(floorNumber)
        self.sortFloorList()       

    def sortFloorList(self):
        if self.direction is str("up"):
            self.floorList.sort()
        else:
            self.floorList.sort(reverse=True)

class floor:
    def __init__(self, floorNumber, buttons):
        self.floorNumber = floorNumber
        self.directionButtons = buttons

class directionButtons:
    directionButtonList = []
    def __init__(self, direction, floorNumber):
        self.direction = direction
        self.floor = floorNumber
      

class column:
    elevators = []
    def __init__(self, elevators, floors):
        self.elevators = elevators
        self.floors = floors
        

class floorButtons:
    floorButtons = []
    def __init__(self, elevator, requestedFloor, activated):
        self.elevator = elevator
        self.requestedFloor = requestedFloor
        self.activated = activated
        
class openDoorButtons:
    def __init__(self, elevator):
        self.elevator = elevator

class closeDoorButtons:
    def __init__(self, elevator):
        self.elevator = elevator

class door: 
    def __init__(self, elevator, status):
        self.elevator = elevator
        self.status = status

class elevatorController:    
    def __init__(self, numberOfFloor, numberOfElevator, elevators):
        self.numberOfFloor = numberOfFloor
        self.numberOfElevator = numberOfElevator
        floors = []
        self.column = column(elevators, floors)

    def requestElevator(self, floorNumber, direction):
        print("requestElevator")
        print("Request for Elevator on floor = " + str(floorNumber))
        print("for direction = " + str(direction))
        elevator = self.findElevator(floorNumber, direction)
        elevator.addToFloorList(floorNumber)
        print("floorlist = " + str(elevator.floorList))
        
        print("REQUESTED elevator found = " + str(elevator.elevatorNumber))
        self.operateElevator(elevator)
        return elevator
        
    def findElevator(self, floorNumber, direction):
        print("findElevator  for floorNumber = " + str(floorNumber))
        print("findElevator  for direction = " + str(direction))
        print("column elevators  = " + str(self.column.elevators))

        for elevator in self.column.elevators: 
            if floorNumber == elevator.floorNumber and elevator.status == "stopped" and elevator.direction == direction:
                print("elevator found is on floor = " + str(elevator.floorNumber))
                return elevator
            if floorNumber == elevator.floorNumber and elevator.status =="idle":
                print("elevator found is on floor = " + str(elevator.floorNumber))
                return elevator
            if direction == "up" and elevator.direction == direction and floorNumber > elevator.floorNumber:
                print("elevator found is on floor = " + str(elevator.floorNumber))
                return elevator
            if direction == "down" and elevator.direction == direction and floorNumber < elevator.floorNumber:
                print("elevator found is on floor = " + str(elevator.floorNumber))
                return elevator
            if floorNumber != elevator.floorNumber and elevator.status == "idle":
                print("elevator found is on floor = " + str(elevator.floorNumber))
                return elevator
                

        print("sending elevator with shortest floorlist")
        return self.shortestFloorList()
        

    def requestFloor(self, elevator, requestedFloor):
        print("requestFloor")
        print("Inside elevator = " + str(elevator.elevatorNumber))
        print("The requested floor is = " + str(requestedFloor))
        elevator.addToFloorList(requestedFloor)
        print("floorlist = " + str(elevator.floorList))
        self.operateElevator(elevator)
    
    def shortestFloorList(self):
        shortestLength = 10
        shortestListElevator = None
        for elevator in self.column.elevators:
            if len(elevator.floorList) < shortestLength:
                shortestLength = len(elevator.floorList)
                shortestListElevator = elevator

        print("elevator with shortest floor list: " + shortestListElevator.elevatorNumber)

        return shortestListElevator

    def operateElevator(self, elevator):
        print("operating elevator " + str(elevator.elevatorNumber))
        if len(elevator.floorList) > 0:
            requestedFloor = elevator.floorList[0]
            if elevator.floorNumber == requestedFloor: 
                print("open doors elevator on floor " + str(elevator.floorNumber))
                self.OpenDoors(elevator)
            elif elevator.floorNumber > requestedFloor:
                print("move down elevator to requested floor " + str(elevator.floorNumber))
                self.moveDown(elevator)
            elif elevator.floorNumber < requestedFloor:
                print("move up elevator to requested floor " + str (elevator.floorNumber))
                self.moveUp(elevator)

        else:
             elevator.status = "idle"
             print("elevator status is " + str(elevator.status))

    def moveDown(self, elevator):
        elevator.status = "moving"
        print("elevator status is " + str(elevator.status))
        requestedFloor = elevator.floorList[0]
        while (elevator.floorNumber > requestedFloor): 
           elevator.floorNumber = elevator.floorNumber - 1
           print("moved down to floor", elevator.floorNumber)
        self.OpenDoors(elevator)
        
    def moveUp(self, elevator):
        elevator.status = "moving"
        print("elevator status is " + str(elevator.status))
        requestedFloor = elevator.floorList[0]
        while (elevator.floorNumber < requestedFloor):
            elevator.floorNumber = elevator.floorNumber + 1
            print("moved up to floor", elevator.floorNumber)
        self.OpenDoors(elevator)

    def OpenDoors(self, elevator): 
        elevator.status = "stopped"
        print("elevator status is " + str(elevator.status))
        print("opening doors for elevator " + str(elevator.elevatorNumber) + " on floor  " + str(elevator.floorNumber))
        time.sleep(5)
        elevator.floorList.remove(elevator.floorList[0])
        print("updated floorList = " + str(elevator.floorList))
        self.CloseDoors(elevator)

    def CloseDoors(self, elevator): 
        elevator.status = "stopped"
        print("elevator status is " + str(elevator.status))
        print("closing doors for elevator " + str(elevator.elevatorNumber))
        time.sleep(5)
        print("doors are closed for elevator "+ str(elevator.elevatorNumber))
        self.operateElevator(elevator)



        
        

        
        
#test 
#elevators = [elevator("idle", "none", 1, "one"), elevator("idle", "none", 4, "two")]
#controller = elevatorController(10, 2, elevators) 
#controller.requestElevator(3, "up")
#elevators[0].floorList = [5,6,9]
#elevators[1].floorList = [7,5]
#controller.requestFloor (elevators[0], 7)
#controller.operateElevator (elevators[1])


#test  
#elevators = [elevator("idle", "none", 1, "one"), elevator("idle", "none", 10, "two")]
#controller = elevatorController(10, 2, elevators) 
#elevator = controller.requestElevator(4, "up")
#if elevator.floorNumber == 4:
#    controller.requestFloor (elevator, 7)
#elevators[0].floorList = [5,6,9]

File: Residential_Controller.cs
using System;
using System.Collections.Generic;

namespace Rocket_Elevators_Controllers
{
    class Program
    {
        static void Main(string[] args)
        {
        
        

        List<Elevator> ElevatorList = new List<Elevator>();
        Elevator elevator1 = new Elevator(1, 1);
        ElevatorList.Add(elevator1);

        ElevatorController Controller = new ElevatorController (10,2); 
        Console.WriteLine("Request elevator");
        Controller.RequestElevator(4, "up");
        Console.WriteLine("Request floor");
        Controller.RequestFloor(ElevatorList [0], 7 );
        
        }
    }

    class Elevator {
        public int floorNumber;
        public string status;
        public string direction; 
        public int elevatorNumber;
        public List<int> floorList;
        public Elevator(int floorNumber, int elevatorNumber)
    {   this.floorNumber = floorNumber; 
        this.status = "idle"; 
        this.direction = "GoingNowhere"; 
        this.elevatorNumber = elevatorNumber; 
        this.floorList = new List<int>();
    }
        public void AddRequestedFloorToFloorList(int requestedFloor){
            this.floorList.Add(requestedFloor);
        }

        public bool EmptyFloorList()
        {
            return floorList.Count==0?true:false;
        } 

    }
    
    class Floor {
        public int floorNumber;
        public List<DirectionButtons> buttons; 
       
        public Floor(int floorNumber, List<DirectionButtons> buttons)
    {   this.floorNumber = floorNumber; 
        this.buttons = buttons;
        }
    }

    class DirectionButtons {
        public string direction; 
        public int floorNumber; 
        public DirectionButtons(string direction, int floorNumber)
    {   this.direction = direction; 
        this.floorNumber = floorNumber;

    }
    }
    class Column {
        public int elevators; 
        public int floors; 
        public Column (int elevators, int floors)
   {    this.elevators = elevators; 
        this.floors = floors; 
   }
   }

   class FloorButtons {
       public int elevators; 
       public int requestedFloor; 
       public string activated; 
       public FloorButtons (int elevators, int requestedFloor, string activated)
    {  this.elevators = elevators; 
       this.requestedFloor = requestedFloor;
       this.activated = activated;
       }
   }
    class OpenDoorButtons {
      public int elevators;
      public OpenDoorButtons (int elevators)
    { this.elevators = elevators; 
      }
    }

    class CloseDoorButtons {
        public int elevaotors;
        public CloseDoorButtons (int elevators)
    { this.elevaotors = elevators;
     }
    }

    class Door {
        public int elevators; 
        public string status; 
        public Door(int elevators, string status)
    { this.elevators = elevators;
      this.status = status;
    }
    }

    class ElevatorController {
        public int numberOfFloor; 
        public int numberOfElevator;
        public List<int> floorList; 
        public List<Elevator> elevatorList;
        public ElevatorController(int numberOfFloor, int numberOfElevator)
    {   this.numberOfFloor = numberOfFloor; 
        this.numberOfElevator = numberOfElevator; 
        this.floorList = new List<int>();

        this.elevatorList = new List<Elevator> ();
           for (int i = 0; i < numberOfElevator; i++) {
               this.elevatorList.Add (new Elevator(1, i));
           }
    }


       public void RequestElevator(int floorNumber, string direction)
       {
           Console.WriteLine("Request Elevator On Floor " + floorNumber.ToString() + ", going " + direction);
        var elevator = FindElevator(floorNumber,  direction);   

          //RequestFloor(elevator, floorNumber); 
          elevator.AddRequestedFloorToFloorList(floorNumber);
          OperateElevator(elevator); 
       }

       public Elevator FindElevator(int floorNumber, string direction)
       { 
           foreach(Elevator elevator in this.elevatorList)
           { 
                if (floorNumber == elevator.floorNumber || elevator.status == "stopped" || elevator.direction == direction)
                {
                    return elevator; 
                }
                else if (floorNumber == elevator.floorNumber || elevator.status == "idle")
                {
                    return elevator; 
                }
                else if (direction == "up" || elevator.direction == direction || floorNumber < elevator.floorNumber)
                {
                    return elevator; 
                }
                else if (direction == "down" || elevator.direction == direction || floorNumber > elevator.floorNumber)
                {
                    return elevator; 
                }
                else if (floorNumber != elevator.floorNumber || elevator.status == "idle")
                {
                    return elevator; 
                }
           }
           return ShortestFloorList();
       }
       public void RequestFloor(Elevator elevator, int requestedFloor)
       {
           elevator.AddRequestedFloorToFloorList(requestedFloor);
           Console.WriteLine("Request For Floor Number " + requestedFloor);
           OperateElevator(elevator); 
       }
        
        public void MoveDown (Elevator elevator, int requestedFloor)
        {
            
                while (elevator.floorNumber > requestedFloor) {
                    elevator.floorNumber--; 
                }; 
            Console.WriteLine("Moved Down to Floor" + requestedFloor.ToString()); 

            OpenDoors(elevator, requestedFloor); 
            
        }

        public void MoveUp (Elevator elevator, int requestedFloor)
        { requestedFloor = elevator.floorList[0]; 
                while (elevator.floorNumber < requestedFloor) {
                    elevator.floorNumber++; 
                }; 
            
            Console.WriteLine ("Moving up to floor " + requestedFloor.ToString());

            OpenDoors(elevator, requestedFloor); 
        }

        public void OpenDoors (Elevator elevator, int requestedFloor)
        {
            elevator.floorList.Remove(requestedFloor);
            this.wait(5000);
        
            Console.WriteLine ("Opening Doors on floor " + requestedFloor.ToString()); 
    
            CloseDoors(elevator, requestedFloor);
        
        }

        public void CloseDoors (Elevator elevator, int requestedFloor)
        {
            this.wait(5000);
            Console.WriteLine ("Closing Doors on floor " + requestedFloor.ToString()); 

            OperateElevator(elevator); 
        }
       public void OperateElevator(Elevator elevator)
       {
           
           if(elevator.EmptyFloorList() == false)
            {
               var requestedFloor = elevator.floorList[0]; 
                if (elevator.floorNumber == requestedFloor)
            { 
                OpenDoors(elevator, requestedFloor); 
                Console.WriteLine ("Opening Doors  " + requestedFloor.ToString()); 
                
            }
            
                 else if (elevator.floorNumber > requestedFloor)
            {
                MoveDown(elevator, requestedFloor); 
                 

            }
                else if (elevator.floorNumber < requestedFloor)
            {   
                MoveUp(elevator, requestedFloor); 
                 
                }
            
           }

             else elevator.status = "idle"; 
        
       }

       public Elevator ShortestFloorList(){
           var ShortestLenght = 10; 
           Elevator ShortestListElevator = null; 
            foreach(Elevator elevator in this.elevatorList)
            { 
                if (elevator.floorList.Count < ShortestLenght) {
                    ShortestLenght = elevator.floorList.Count;
                    ShortestListElevator = elevator; 
                }
            }
            return ShortestListElevator; 
        }
            
        public List<int> SortedFloorList(List<int> floorList, string direction)
        {
               if (direction == "up") 
               {
                floorList.Sort(); 
               }
                else
                {
                floorList.Sort((a, b) => -1* a.CompareTo(b));
                }
               return floorList;

        }
            

      
    public void wait (int ms) {
        System.Threading.Thread.Sleep (ms);
      }
    }


    }

